File: app/src/main/java/ru/vtb/ppcd/example/utils/ExchangeUtils.java
package ru.vtb.ppcd.example.utils;

import ru.vtb.ppcd.example.processor.ExchangeContext;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.SneakyThrows;
import lombok.experimental.UtilityClass;
import org.apache.camel.Exchange;

@UtilityClass
public class ExchangeUtils {

    @SneakyThrows
    public static <T> ExchangeContext<T> toExchangeContext(
            Exchange exchange,
            ObjectMapper objectMapper,
            Class<T> processClass
    ) {
        var incomeMessage = exchange.getMessage().getBody().toString();
        var incomeHeaders = exchange.getIn().getHeaders();
        var messageId = exchange.getIn().getMessageId();

        return ExchangeContext.<T>builder()
                .headers(incomeHeaders)
                .messageId(messageId)
                .message(objectMapper.readValue(incomeMessage, processClass))
                .build();
    }

    @SneakyThrows
    public static <T> void updateExchange(
            Exchange exchange,
            ObjectMapper objectMapper,
            ExchangeContext<T> exchangeContext
    ) {
        var outcomeMessage = objectMapper.writeValueAsString(exchangeContext.getMessage());
        exchange.getMessage().setBody(outcomeMessage);

        var incomeHeaders = exchange.getIn().getHeaders();
        incomeHeaders.forEach((s, o) -> exchange.getIn().setHeader(s, o));
    }
}

File: ms-ppcd-example-app/src/main/kotlin/ru/vtb/ppcd/example/utils/ExchangeUtils.kt
package org.example.ru.vtb.ppcd.example.utils

import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.camel.Exchange
import org.example.ru.vtb.ppcd.example.model.ExchangeContext

class ExchangeUtils {

    companion object {
        fun <T> toExchangeContext(
            exchange: Exchange,
            objectMapper: ObjectMapper,
            processClass: Class<T>?
        ): ExchangeContext<T> {
            val incomeMessage = exchange.message.body.toString()
            val incomeHeaders = exchange.getIn().headers
            val messageId = exchange.getIn().messageId

            return ExchangeContext(
                message = objectMapper.readValue(incomeMessage, processClass),
                messageId = messageId,
                headers = incomeHeaders
            )
        }

        fun <T> updateExchange(
            exchange: Exchange,
            objectMapper: ObjectMapper,
            exchangeContext: ExchangeContext<T>
        ) {
            objectMapper.writeValueAsString(exchangeContext.message)
                .let { outcomeMessage ->
                    exchange.message.body = outcomeMessage
                }

            val incomeHeaders = exchange.`in`.headers
            incomeHeaders.forEach { (key: String, value: Any) ->
                exchange.`in`.setHeader(key, value)
            }
        }
    }
}